### ISSUE TITLE
ASDF interface for Table

### ISSUE THREAD

**Issue by astrofrog** [astropy/astropy#3733]
We should add a reader/writer for ASDF for tables (which can have pyasdf as an optional dependency for now). But first, a philosophical question: should we store the table by rows or by columns? For the former, we'd use a structured array, while for the latter, we'd use a dictionary of Numpy arrays. The latter would be faster because that is how Table is stored internally, and it also allows us to store each column's metadata alongside the column it applies to. So the difference would be:

``` python
t = Table(...)

# Row-first:

tree1 = {'data':np.array(t)}

# also store the column metadata in another structure

ff = AsdfFile(tree1)
ff.write_to("example1.asdf")

# Column-first:

tree2 = {}
for colname in t.columns:
    tree2[colname] = {}
    tree2[colname]['data'] = t.columns[colname]
    tree2[colname]['unit'] = str(t.columns[colname].unit)
    tree2[colname]['description'] = str(t.columns[colname].description)

ff = AsdfFile(tree2)
ff.write_to("example2.asdf")
```

The latter seems more elegant to me and is a bit faster in real-life tests, and I like the inclusion of the column meta-data alongside each array. This looks like:

```
--- !core/asdf
col1:
  data: !core/ndarray
    source: 2
    datatype: [ascii, 25]
    byteorder: big
    shape: [4455]
  description: GLIMPSE source name
  unit: '---'
col10:
  data: !core/ndarray
    source: 9
    datatype: [ascii, 1]
    byteorder: big
    shape: [4455]
  description: "Variablity flag (2) \n    Y = The magnitudes for the two epochs differ
    by at least 0.3 mag at either \n        (or both) 4.5 microns or 8.0 microns.
    \n    N = Not variable."
  unit: '---'
col2:
  data: !core/ndarray
    source: 8
    datatype: float32
    byteorder: little
    shape: [4455]
  description: "Galatic longitude (1) These coordinates are set to the average position
    of the source at \n          4.5 and at 8.0 microns, as in Table 2."
  unit: deg
```

cc @taldcroft @embray @taldcroft

**Comment by mdboom** [comment]
Hmm...  I'm not crazy about the "dictionary of columns" approach, as it creates an asymmetry between binary tables and YAML tables.  One of the design features here as been that binary tables and YAML tables are stored in as similar a way as possible so that converting from one to another is trivial (particularly from the perspective on a non-Python application).  Also, we'd at a minimum want ordered dictionaries to maintain column order, particularly if these would be converted to binary tables/C structs and back.

This does illustrate, however, the need for column-specific metadata which ASDF currently doesn't have.

**Comment by embray** [comment]
I'm not sure I exactly would want this approach, but am somewhat in favor of allowing table storage either by row or by column.  I've suggested this in the past for example, in the context of exploding/recombining ASDF files, and it would be really nice to be able to do this with table columns (this is essentially how CASA tables work as I understand it).

But that's an issue for ASDF and maybe not in the first version.

**Comment by mdboom** [comment]
It strikes me that there's two things we want to support here:

1) Column-specific metadata
2) Column-oriented binary data

I think the easiest/cleanest/most flexible way to support this is this:  Add a new type called "table" which takes a list of arrays, each array being a column.  For column-oriented tables, the arrays would each point to their own block.  For row-oriented tables, it will probably be special striding on a single binary block.

Having done that, in the interest of "only one way to do it", we may want to remove the current support for what are essentially "record arrays".  This would also simplify the specification and the code somewhat.  But maybe there's a use case for "record arrays" vs. what I'm proposing that I'm missing (but I doubt it since what I'm describing should be inclusive of anything a record array can do).

**Comment by mdboom** [comment]
I think I've answered my own question about the record arrays -- they are definitely useful to retain -- particularly for, say, a 2D RGB image.  Tables are, at least as astropy implements them, one-dimensional.  So anything I add here is probably just in addition to what already exists.

**Comment by embray** [comment]
Yeah, that's a little tricky because you can model an RGB image as a "table".  But I think I like the "record array" concept as a lower-level object with "table" being a higher-level object with additional semantics (such as being a 1D array of rows), which *may* be stored as a record-array, or as the more column-oriented approach.

**Comment by embray** [comment]
And to toss out another example, a record array is a better structure for storing any sort of N-d vector field.

**Comment by embray** [comment]
One more point--back when this was called FINF and was first thinking about it, I didn't think about adding a separate table type, since there was no difference between a table and a 1-D structured array.  Though I felt that if we did want to add a type specifically for tables it would just be a subclass of the array type and would have additional constraints (would have to be 1D, would have to have a struct-like dtype, etc.).

**Comment by astrofrog** [comment]
> And to toss out another example, a record array is a better structure for storing any sort of N-d vector field.

Unless the vectors are the same length for all rows, right?

**Comment by embray** [comment]
I'm not sure I understand.  I meant like a lattice with vector values at each point.

**Comment by pllim** [comment]
c/c @drdavella

**Comment by drdavella** [comment]
Closing since this has been resolved by #8261. There's still WIP to enable mixin columns, but the general idea here is in place.
